#' Read plink 12 coding file (plink --recode 12 )into one-hot encoding or additive modle matrix.
#'
#' @author Byungju Kim (bjk@inu.ac.kr)
#' @param ped a 1/2 coded plnk pedigree file (+7 metadata column)
#' @param map plink genomic map file which describe genomic coordinate and unique SNP IDs.
#' @param onehot choice of one-hot encoding or additive model.
#' @examples
#' ped2Add(ped,map)
#'

read12toAdd <- function(ped,map,onehot=F)
{
  ## pedToAdd        : convert PED file into one-hot or additive model. ped file should be generated by --recode 12 option.
  ## a.k.a. pedTo012 : convert Ped To 012 genotype code where 0 for A2_homozygosity, 1 for Heterozygosity and 2 for A1_homozygosity

  tmp.ped<-read.table(ped,header=F,colClasses=c("character"))
  tmp.map<-read.table(map,header=F)[,2]
  tmp.map<-as.character(tmp.map)
  sample<-tmp.ped$V2

  df1<-tmp.ped[,7:dim(tmp.ped)[2]]
  df1<-as.matrix(df1)
  d1<-df1[,c(T,F)];d2<-df1[,c(F,T)]

  df1<-matrix(paste(d1,d2,sep=""),ncol=length(tmp.map))
  colnames(df1)<-tmp.map

  rm(d1,d2)
  df1<-as_tibble(df1)
  df1$sample<-tmp.ped$V2
  df1<-df1 %>% group_by(sample) %>% gather_("rsid","genotype",as.character(tmp.map)) %>% mutate(genotype=ifelse(substr(genotype,1,1) < substr(genotype,2,2),genotype,stringi::stri_reverse(genotype)))

  if(onehot==F) {
    df1$genotype2<-str_count(df1$genotype,"1")
    df1<-df1 %>% select(-genotype) %>% spread(rsid,genotype2) %>% select(sample,as.character(tmp.map)) %>% ungroup %>% select(-sample)
  } else {

    df1$genotype2<-1
    df1<-df1 %>% unite(rsid2,rsid,genotype,sep="_") %>% spread(rsid2,genotype2) %>% ungroup %>% select(-sample)
    df1[is.na(df1)]<-0
  }
  df1<-as.matrix(df1)
  row.names(df1) <- sample

  return(df1)
}

readADtoAdd <-funciton(ped,map,onehot=F)
{
  tmp.ped<-read.table(ped,header=F,colClasses=c("character"))
  tmp.map<-read.table(map,header=F)[,2]
  tmp.map<-as.character(tmp.map)
  sample<-tmp.ped$V2
  df1<-tmp.ped[,7:dim(tmp.ped)[2]]
  df1<-as.matrix(df1)
}
